spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SoftMax.scala

```diff
@@ -31,61 +31,74 @@ import scala.reflect.ClassTag
  * where shift = max_i(x_i).
  */
 @SerialVersionUID(- 7842335603491194236L)
-class SoftMax[T: ClassTag]()(implicit ev: TensorNumeric[T]) extends TensorModule[T] {
+class SoftMax[T: ClassTag](var pos: Int = 1)(implicit ev: TensorNumeric[T])
+  extends TensorModule[T] {
 
   @transient
   private var results: Array[Future[Unit]] = null
 
+  private def getPositiveDimension(input: Tensor[T]): Int = {
+    val inputDim = input.nDimension() // data batch dim
+    pos = if (pos <= 0) {
+      inputDim + pos
+    }
+    else pos
+    require(1 <= pos && pos <= input.nDimension(),
+      s"Invalid position: $pos ." + s"input dimension ${input.nDimension()}")
+    pos
+  }
+
   override def updateOutput(input: Tensor[T]): Tensor[T] = {
     require(1 <= input.nDimension() && input.nDimension() <= 4,
       "1D, 2D, 3D or 4D tensor expected" +
         s"input dimension ${input.nDimension()}")
-    val (nFrame, stride) = if (input.nDimension() == 1) {
+    pos = getPositiveDimension(input)
+
+    val (nFrame, stride) = if (input.nDimension() - pos == 0) {
       (1, 1)
-    } else if (input.nDimension() == 2) {
-      (input.size(1), 1)
-    } else if (input.nDimension() == 3) {
-      (1, input.size(2) * input.size(3))
+    } else if (input.nDimension() - pos == 1) {
+      (input.size(pos), 1)
+    } else if (input.nDimension() - pos == 2) {
+      (1, input.size(pos + 1) * input.size(pos + 2))
     } else {
-      (input.size(1), input.size(3) * input.size(4))
+      (input.size(pos), input.size(pos + 2) * input.size(pos + 3))
     }
     if (results == null || results.length != nFrame * stride) {
       results = new Array[Future[Unit]](nFrame * stride)
     }
     output.resizeAs(input)
-    SoftMax.updateOutput[T](input, output, results)
+    SoftMax.updateOutput[T](input, output, results, pos)
     output
   }
-
   override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
+    pos = getPositiveDimension(input)
```

Comment:
pos should be computed only once, not every time we do updates.

Reply:
ok, I am not sure whether the updateGradInput will always be called after updateOutput, so I added getPositiveDimension in both of them.